### PR TITLE
Rxjava 2

### PR DESCRIPTION
**Pull request by piotr-sumo**

### CODE REVIEW

**Comment by piotr-sumo** [comment]
Dear reviewers, I need some help with this build. It fails while locally I can assemble the project.
For example, `./gradlew :instrumentation:rabbitmq-2.7:javaagent:test` fails locally with `com.github.dockerjava.api.exception.NotFoundException: Status 404: {"message":"No such image: testcontainers/ryuk:0.3.0"}` which is not related to my changes.
The same is with `:instrumentation:apache-camel-2.20:javaagent:test`.

**Comment by jkwatson** [comment]
@iNikem @trask @anuraaga any ideas on helping @piotr-sumo here?

**Comment by iNikem** [comment]
@piotr-sumo Can you run tests in the main branch of this repo? I mean, is this problem of this PR or your environment in general?

**Comment by piotr-sumo** [comment]
@iNikem I will run tests on `main` and report update here.

**Comment by piotr-sumo** [comment]
@iNikem I've run tests from fresh master on jdk 15 and it fails unfortunately.
```RmiTest > Client call creates spans FAILED
    java.rmi.ConnectException: Connection refused to host: 52.6.177.231; nested exception is:
        java.net.ConnectException: Operation timed out```

**Comment by anuraaga** [comment]
@piotr-sumo Do you mind pasting the larger stack trace there?

**Comment by piotr-sumo** [comment]
@anuraaga this is one of the failing tests:
```
> Task :instrumentation:rmi:javaagent:test

RmiTest > Client call creates spans FAILED
    java.rmi.ConnectException: Connection refused to host: 54.159.152.5; nested exception is:
        java.net.ConnectException: Operation timed out
        at java.rmi/sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:623)
        at java.rmi/sun.rmi.transport.tcp.TCPChannel.createConnection(TCPChannel.java:209)
        at java.rmi/sun.rmi.transport.tcp.TCPChannel.newConnection(TCPChannel.java:196)
        at java.rmi/sun.rmi.server.UnicastRef.invoke(UnicastRef.java:132)
        at java.rmi/java.rmi.server.RemoteObjectInvocationHandler.invokeRemoteMethod(RemoteObjectInvocationHandler.java:217)
        at java.rmi/java.rmi.server.RemoteObjectInvocationHandler.invoke(RemoteObjectInvocationHandler.java:162)
        at RmiTest.Client call creates spans_closure1(RmiTest.groovy:37)
        at RmiTest.Client call creates spans_closure1(RmiTest.groovy)
        at io.opentelemetry.instrumentation.test.utils.TraceUtils.runUnderTrace_closure1(TraceUtils.groovy:32)
        at groovy.lang.Closure.call(Closure.java:405)
        at groovy.lang.Closure.call(Closure.java:421)
        at io.opentelemetry.instrumentation.test.utils.TraceUtils.runUnderTrace(TraceUtils.groovy:31)
        at RmiTest.Client call creates spans(RmiTest.groovy:35)

        Caused by:
        java.net.ConnectException: Operation timed out
            at java.base/sun.nio.ch.Net.connect(Net.java:574)
            at java.base/sun.nio.ch.Net.connect(Net.java:563)
            at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:588)
            at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:333)
            at java.base/java.net.Socket.connect(Socket.java:648)
            at java.base/java.net.Socket.connect(Socket.java:597)
            at java.base/java.net.Socket.<init>(Socket.java:520)
            at java.base/java.net.Socket.<init>(Socket.java:294)
            at java.rmi/sun.rmi.transport.tcp.TCPDirectSocketFactory.createSocket(TCPDirectSocketFactory.java:40)
            at java.rmi/sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:617)
            ... 12 more
```

**Comment by mateuszrzeszutek** [comment]
@piotr-sumo Are you using any VPN or proxy server?

**Comment by piotr-sumo** [comment]
@mateuszrzeszutek Yes, I have corporate VPN turned on. Let me run those tests with VPN disabled.

**Comment by piotr-sumo** [comment]
@mateuszrzeszutek withouth VPN I get
```> Task :instrumentation:spymemcached-2.12:javaagent:test FAILED

SpymemcachedTest > initializationError FAILED
    org.testcontainers.containers.ContainerLaunchException: Container startup failed
        at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:331)
        at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:312)
        at SpymemcachedTest.setupSpec(SpymemcachedTest.groovy:51)

        Caused by:
        org.testcontainers.containers.ContainerFetchException: Can't get Docker image: RemoteDockerImage(imageName=memcached:latest, imagePullPolicy=DefaultPullPolicy())
            at org.testcontainers.containers.GenericContainer.getDockerImageName(GenericContainer.java:1282)
            at org.testcontainers.containers.GenericContainer.logger(GenericContainer.java:616)
            at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:321)
            ... 2 more

            Caused by:
            com.github.dockerjava.api.exception.NotFoundException: Status 404: {"message":"No such image: testcontainers/ryuk:0.3.0"}
                at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.execute(DefaultInvocationBuilder.java:241)
                at org.testcontainers.shaded.com.github.dockerjava.core.DefaultInvocationBuilder.post(DefaultInvocationBuilder.java:125)
                at org.testcontainers.shaded.com.github.dockerjava.core.exec.CreateContainerCmdExec.execute(CreateContainerCmdExec.java:33)
                at org.testcontainers.shaded.com.github.dockerjava.core.exec.CreateContainerCmdExec.execute(CreateContainerCmdExec.java:13)
                at org.testcontainers.shaded.com.github.dockerjava.core.exec.AbstrSyncDockerCmdExec.exec(AbstrSyncDockerCmdExec.java:21)
                at org.testcontainers.shaded.com.github.dockerjava.core.command.AbstrDockerCmd.exec(AbstrDockerCmd.java:35)
                at org.testcontainers.shaded.com.github.dockerjava.core.command.CreateContainerCmdImpl.exec(CreateContainerCmdImpl.java:595)
                at org.testcontainers.utility.ResourceReaper.start(ResourceReaper.java:89)
                at org.testcontainers.DockerClientFactory.client(DockerClientFactory.java:201)
                at org.testcontainers.LazyDockerClient.getDockerClient(LazyDockerClient.java:14)
                at org.testcontainers.LazyDockerClient.listImagesCmd(LazyDockerClient.java:12)
                at org.testcontainers.images.LocalImagesCache.maybeInitCache(LocalImagesCache.java:68)
                at org.testcontainers.images.LocalImagesCache.get(LocalImagesCache.java:32)
                at org.testcontainers.images.AbstractImagePullPolicy.shouldPull(AbstractImagePullPolicy.java:18)
                at org.testcontainers.images.RemoteDockerImage.resolve(RemoteDockerImage.java:65)
                at org.testcontainers.images.RemoteDockerImage.resolve(RemoteDockerImage.java:26)
                at org.testcontainers.utility.LazyFuture.getResolvedValue(LazyFuture.java:17)
                at org.testcontainers.utility.LazyFuture.get(LazyFuture.java:39)
                at org.testcontainers.containers.GenericContainer.getDockerImageName(GenericContainer.java:1280)
                ... 4 more

1 test completed, 1 failed
```

**Comment by iNikem** [comment]
Do you have any local Docker setup/conf that could prevent from pulling images from Docker Hub?

**Comment by piotr-sumo** [comment]
I came across https://github.com/testcontainers/testcontainers-java/issues/3574 and pulled that image manually. I've also merged latest `main`. Let's see what happens.
@iNikem I don't have such a docker config. Could you provide any or point me to the right direction?

**Comment by iNikem** [comment]
> I came across [testcontainers/testcontainers-java#3574](https://github.com/testcontainers/testcontainers-java/issues/3574) and pulled that image manually. I've also merged latest `main`. Let's see what happens.
> @iNikem I don't have such a docker config. Could you provide any or point me to the right direction?

A, if this testcontainers bug (or we use too old version) and your local env works after pulling image manually, then problem solved, right?

**Comment by piotr-sumo** [comment]
@iNikem it is almost solved. I get 
Screenshot 2021-02-01 at 15 12 44
but when I open the file in Chrome I can see
Screenshot 2021-02-01 at 15 13 01

File: instrumentation/rxjava-2.0/javaagent/rxjava-2.0-javaagent.gradle [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  pass {
+    group = "io.reactivex.rxjava2"
+    module = "rxjava"
+    versions = "[2.0.0,)"
+  }
+}
+
+dependencies {
+  implementation group: 'org.reactivestreams', name: 'reactive-streams', version: '1.0.0'
+  library group: 'io.reactivex.rxjava2', name: 'rxjava', version: '2.0.0'
```
Comment (anuraaga) [inline instrumentation/rxjava-2.0/javaagent/rxjava-2.0-javaagent.gradle:13]: Does this instrumentation use rxjava, it seems to only instrument reactive streams? I couldn't find any usage outside of tests.  If so we can name the instrumentation reactivestreams (we'd probably want to add tests with reactor too). Also I wonder how this instrumentation interops with other reactivestreams instrumentation we have, such as `reactor`.

Also for this instrumentation especially, I think it would help to start with library instrumentation instead of javaagent instrumentation - It uses so much bytecode instrumentation I don't think we could extract a library for it. An example library that instruments rxjava2 for context propagation is https://github.com/line/armeria/tree/master/rxjava2

Comment (piotr-sumo) [reply instrumentation/rxjava-2.0/javaagent/rxjava-2.0-javaagent.gradle:13]: I was porting DD instrumentation as mentioned here https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1866. And the DD instrumentation uses javaagent instrumentation (https://github.com/DataDog/dd-trace-java/tree/master/dd-java-agent/instrumentation/rxjava-2)

Comment (anuraaga) [reply instrumentation/rxjava-2.0/javaagent/rxjava-2.0-javaagent.gradle:13]: @piotr-sumo Would you be able to help us by extracting this into library instrumentation? Going forward we do want to make sure we have it where we can. If it's too much for now though, I think this time we can stick with this and refactor (a fair amount I guess) later.

Comment (piotr-sumo) [reply instrumentation/rxjava-2.0/javaagent/rxjava-2.0-javaagent.gradle:13]: @anuraaga Sure, I'd like to contribute and extract this into library instrumentation. First, I'll familiarise myself with https://github.com/line/armeria/tree/master/rxjava2 and proceed with library instrumentation for `rxjava-2`.

**Comment by piotr-sumo** [comment]
RxJava2 library instrumentation PR can be found here https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/2191

File: instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/CompletableInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+              .and(named("subscribe"))
+              .and(takesArguments(1))
+              .and(takesArgument(0, named("io.reactivex.CompletableObserver"))),
+          PropagateParentSpanAdvice.class.getName());
+      return transformers;
+    }
+  }
+
+  public static class CaptureParentSpanAdvice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
```
Comment (aimiebell) [inline instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/CompletableInstrumentation.java:63]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/FlowableInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+              .and(named("subscribe"))
+              .and(takesArguments(1))
+              .and(takesArgument(0, named("org.reactivestreams.Subscriber"))),
+          PropagateParentSpanAdvice.class.getName());
+      return transformers;
+    }
+  }
+
+  public static class CaptureParentSpanAdvice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
```
Comment (aimiebell) [inline instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/FlowableInstrumentation.java:63]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/MaybeInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+              .and(named("subscribe"))
+              .and(takesArguments(1))
+              .and(takesArgument(0, named("io.reactivex.MaybeObserver"))),
+          PropagateParentSpanAdvice.class.getName());
+      return transformers;
+    }
+  }
+
+  public static class CaptureParentSpanAdvice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
```
Comment (aimiebell) [inline instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/MaybeInstrumentation.java:63]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/ObservableInstrumentation.java [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+              .and(named("subscribe"))
+              .and(takesArguments(1))
+              .and(takesArgument(0, named("io.reactivex.Observer"))),
+          PropagateParentSpanAdvice.class.getName());
+      return transformers;
+    }
+  }
+
+  public static class CaptureParentSpanAdvice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
```
Comment (aimiebell) [inline instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/ObservableInstrumentation.java:63]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/SingleInstrumentation.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    }
+  }
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return Collections.singletonList(new Instrumentation());
+  }
+
+  public static class CaptureParentSpanAdvice {
+    @Advice.OnMethodExit(suppress = Throwable.class)
```
Comment (aimiebell) [inline instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/SingleInstrumentation.java:63]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

**Comment by iNikem** [comment]
@piotr-sumo just to check, do you plan to finish this PR now when #2191 is merged?

**Comment by piotr-sumo** [comment]
@iNikem yes, I'm working on this now

**Comment by piotr-sumo** [comment]
@iNikem @mateuszrzeszutek @anuraaga could you review my changes?

**Comment by anuraaga** [comment]
@piotr-sumo Can you rework this to use the library instrumentation? We like the agent to just register library instrumentation whenever it can and it will make this a lot smaller. I think we can instrument `RxJavaPlugins` to enable our assembly - ideally we could instrument the static initializer of the class though I'm not sure I've seen us use that pattern. @trask do you know a good pattern for this?

**Comment by piotr-sumo** [comment]
@anuraaga the library instrumentation is already merged https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/2191
After reading the last comment in https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/1866 I've though that the javaagent instrumentation is still required.

However, I'm happy to contribute and I will rework implementation in this PR to meet your requirements.

**Comment by anuraaga** [comment]
@piotr-sumo Yup, it's great we have library instrumentation since now we can use it from this javaagent instrumentation :) We do need javaagent instrumentation but it should be implementable with a small amount of code that just initializes the library instrumentation instead of reimplementing it.

We use a similar pattern for reactor

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/reactor-3.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/reactor/ReactorHooksAdvice.java#L14

and it actually answered my question on how to run code post-static initializer :D

**Comment by piotr-sumo** [comment]
@anuraaga thanks for the hint! I will apply it!

**Comment by piotr-sumo** [comment]
@anuraaga I've applied your suggestion.

File: testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/asserts/TraceAssert.groovy [hunk 7]
```diff
@@ -11,6 +11,7 @@ import groovy.transform.stc.ClosureParams
 import groovy.transform.stc.SimpleType
 import io.opentelemetry.instrumentation.testing.util.TelemetryDataUtil
 import io.opentelemetry.sdk.trace.data.SpanData
+
```
Comment (anuraaga) [inline testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/asserts/TraceAssert.groovy:14]: Can you revert this change?

**Review by anuraaga — APPROVED** [review]
Thanks!

File: testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/asserts/TraceAssert.groovy (hunk 7 above)
Comment (piotr-sumo) [reply testing-common/src/main/groovy/io/opentelemetry/instrumentation/test/asserts/TraceAssert.groovy:14]: :+1

**Comment by piotr-sumo** [comment]
@anuraaga @mateuszrzeszutek could you re-run the checks? It failed on part that is not related to my commit. Unfortunately I don't have permissions to trigger re-run myself.

File: instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/RxJava2InstrumentationModule.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  public static class PluginInstrumentation implements TypeInstrumentation {
+
+    @Override
+    public ElementMatcher<? super TypeDescription> typeMatcher() {
+      return named("io.reactivex.plugins.RxJavaPlugins");
+    }
+
+    @Override
+    public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+      return Collections.singletonMap(isTypeInitializer(), RxJavaPluginsAdvice.class.getName());
```
Comment (mateuszrzeszutek) [inline instrumentation/rxjava-2.0/javaagent/src/main/java/io/opentelemetry/instrumentation/rxjava2/RxJava2InstrumentationModule.java:41]: This might sound weird, but can you hard-code the class name instead?
```suggestion
      return Collections.singletonMap(isTypeInitializer(), RxJava2InstrumentationModule.class.getName() + "$RxJavaPluginsAdvice");
```
[This doc](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/writing-instrumentation.md#why-hard-code-advice-class-names) from our contributing folder should explain why it is preferable to hard-code advice class names (although it does look a bit werid).

**Review by mateuszrzeszutek — APPROVED** [review]
LGTM 👍

**Review by mateuszrzeszutek — APPROVED** [review]
